### PR TITLE
[xla:gpu] Add e2e tests for DynamicSliceFusionV2Thunk + ThunkEmitter support

### PR DESCRIPTION
**Pull request by ezhulenev**
Plumb `DynamicSliceFusionV2Thunk` into `ThunkEmitter` and fix minor errors in buffer slice verification.

### CODE REVIEW

**Comment by ezhulenev** [comment]
Most of the lines in this PR are in test, it tests A LOT of different edge cases, including proper clamping of OOB indices. Plus a minor correctness fixes that I found because of unit tests.

File: xla/backends/gpu/tests/dynamic_slice_fusion_v2_test.cc [hunk 1]
```diff
@@ -0,0 +1,1304 @@
...
+#include "xla/tests/literal_test_util.h"
+#include "xla/tsl/platform/test.h"
+#include "xla/xla.pb.h"
+#include "xla/xla_data.pb.h"
+
+namespace xla::gpu {
+namespace {
+
+#if TENSORFLOW_USE_ROCM
+static constexpr char kPlatform[] = "ROCM";
```
Comment (akuegel) [inline xla/backends/gpu/tests/dynamic_slice_fusion_v2_test.cc]: Can we use `CanonicalPlatformName("gpu")` instead?

File: xla/xla.proto [hunk 2]
```diff
@@ -634,6 +634,10 @@ message DebugOptions {
   // dynamic-update-slice operations around library calls.
   optional bool xla_gpu_enable_dynamic_slice_fusion = 105;
 
+  // When true, DynamicSliceFusionV2Thunk verifies at runtime that annotated
+  // offsets match actual XLA-computed DS/DUS offsets. For debugging only.
+  optional bool xla_gpu_dynamic_slice_fusion_verify_offsets = 480;
```
Comment (akuegel) [inline xla/xla.proto]: Please rename to `xla_gpu_experimental_dynamic_slice_fusion_verify_offsets`

Given this is for debugging only, it should be clear from the name that this is not a flag that is meant to be used in production.

Also, please sort it into lexicographic order within the GPU related flags.

File: xla/backends/gpu/tests/dynamic_slice_fusion_v2_test.cc (hunk 1 above)
Comment (ezhulenev) [reply xla/backends/gpu/tests/dynamic_slice_fusion_v2_test.cc]: Done

File: xla/xla.proto (hunk 2 above)
Comment (ezhulenev) [reply xla/xla.proto]: Done

File: xla/xla.proto [hunk 3]
```diff
@@ -778,6 +774,10 @@ message DebugOptions {
   // Experimentally disables binary libraries in GPU compiler passes.
   optional bool xla_gpu_experimental_disable_binary_libraries = 329;
 
+  // When true, DynamicSliceFusionV2Thunk verifies at runtime that annotated
+  // offsets match actual XLA-computed DS/DUS offsets. For debugging only.
+  optional bool xla_gpu_experimental_dynamic_slice_fusion_verify_offsets = 481;
```
Comment (akuegel) [inline xla/xla.proto]: nit: this should be after `xla_gpu_experimental_dump_gpu_executable`

Comment (ezhulenev) [reply xla/xla.proto]: Done

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/BUILD [hunk 4]
```diff
@@ -590,6 +592,7 @@ cc_library(
         "//xla/tsl/platform:statusor",
         "//xla/tsl/protobuf:dnn_proto_cc",
         "@com_google_absl//absl/base:nullability",
+        "@com_google_absl//absl/cleanup",
         "@com_google_absl//absl/container:flat_hash_map",
         "@com_google_absl//absl/container:flat_hash_set",
```
Comment (akuegel) [inline xla/service/gpu/BUILD]: buildcleaner complains that this dependency is not removed (you are removing the corresponding header include in this PR)

**Review by ermilovmaxim — APPROVED** [review]
